bluefog/common/nccl_controller.cc

```diff
@@ -788,6 +776,175 @@ void NCCLController::NeighborAllreduce(TensorTableEntry& entry) {
 #endif
 }
 
+void NCCLController::Allreduce(std::vector<TensorTableEntry>& entries) {
+  auto& first_entry = entries[0];
+  with_device device_guard(first_entry.device);
+
+  void* buffer_data;
+  size_t buffer_len = 0;
+  int64_t num_elements = 0;
+  for (auto& e : entries) {
+    num_elements += e.tensor->shape().num_elements();
+  }
+  std::queue<std::pair<std::string, cudaEvent_t>> event_queue;
+
+  // TODO(ybc) Timeline add record event to measure the time on GPU.
+  const void* fused_input_data;
+  MemcpyInFusionBuffer(entries, fused_input_data, buffer_data, buffer_len);
+  if (timeline_ptr_->Initialized()) {
+    RecordEvent(event_queue, "MEM_CPY_IN");
+  }
```

Comment:
I wonder which of the following records right timeline?
MemcpyInFusionBuffer(entries, fused_input_data, buffer_data, buffer_len);
if (timeline_ptr_->Initialized()) {
  RecordEvent(event_queue, "MEM_CPY_IN");
}
or
if (timeline_ptr_->Initialized()) {
  RecordEvent(event_queue, "MEM_CPY_IN");
}
MemcpyInFusionBuffer(entries, fused_input_data, buffer_data, buffer_len);

Reply:
No difference because all these is async function. The actual time happens in the wait for events. Check that function again, which is very important to understand timeline.